File: frontend/src/app/person.ts
/**
 * Die Angaben eines Menschen — und wie EINE davon herausgeht.
 *
 * <b>Das Einzige auf dieser Plattform, was einem MENSCHEN gehört und keinem
 * Bereich.</b> Sie liegen an seiner Rolle, einzeln versiegelt unter seinem
 * Rollenschlüssel. Keine Pfarrei, keine Schule und kein Verein besitzt sie.
 *
 * <b>Freigeben heisst: neu versiegeln, unter dem Schlüssel des Empfängers.</b>
 * Nicht verweisen — ein Verweis hiesse, dass der Bereich beim Lesen an den
 * Schlüssel des Menschen müsste, und dann wäre die Freigabe ein Versprechen
 * statt einer Massnahme.
 *
 * <b>Jedes Feld ein eigenes Etikett.</b> Trügen zwei dasselbe, ginge der
 * Geheimtext der Telefonnummer am Platz des Geburtstags auf — und eine Freigabe
 * wäre nicht mehr die Freigabe EINER Angabe.
 */

import { loadPublicKey, myEpochKeys } from './area';
import { aad, Field, fromBase64Url, openText, sealText, toBase64Url, type FieldName } from './crypto';
import type { Ring } from './keys';
import { call } from './session';

/** Dieselben sieben wie `ck_person_value_field` und `Person.Fields`. */
export const PERSON_FIELDS =
  ['given_name', 'surname', 'phone', 'born', 'address', 'email', 'nickname'] as const;
export type PersonField = (typeof PERSON_FIELDS)[number];

export const FIELD_LABEL: Record<PersonField, string> = {
  given_name: 'Imię',
  surname: 'Nazwisko',
  phone: 'Telefon',
  born: 'Data urodzenia',
  address: 'Adres',
  email: 'E-mail',
  nickname: 'Przezwisko'
};

/**
 * Die Zuordnung Spaltenname → Etikett.
 *
 * Zwei Namensordnungen für dieselbe Angabe: `app.person_value.field` heisst
 * `given_name`, das Etikett im Kernel heisst auch so — aber `phone` heisst dort
 * `person_phone`. Diese Tabelle ist die einzige Stelle, an der beide
 * zusammenkommen; sie zu umgehen hiesse, sie an einer zweiten Stelle zu raten.
 */
const LABELS: Record<PersonField, FieldName> = {
  given_name: Field.PersonGivenName,
  surname: Field.PersonSurname,
  phone: Field.PersonPhone,
  born: Field.PersonBorn,
  address: Field.PersonAddress,
  email: Field.PersonEmail,
  nickname: Field.PersonNickname
};

/*
 * Das ORIGINAL und die ABSCHRIFT tragen verschiedene Objektarten.
 *
 * Sonst liesse sich eine freigegebene Abschrift an den Platz des Originals
 * schieben: beide nennen dieselbe Rolle und dasselbe Feld, und die Hülle ginge
 * auf. Ein Wort Unterschied schliesst das aus.
 */
export const valueAad = (roleId: string, field: PersonField) =>
  aad('person', 'value', roleId, LABELS[field], 1);

export const releaseAad = (roleId: string, field: PersonField) =>
  aad('person', 'release', roleId, LABELS[field], 1);

/* -- Lesen ------------------------------------------------------------------ */

export interface SealedValue {
  readonly field: PersonField;
  readonly sealed: string;
  readonly updatedAt: string;
}

export interface Release {
  readonly field: PersonField;
  readonly areaId: string;
  readonly areaName: string;
  readonly epoch: number;
  readonly releasedAt: string;
}

export interface Mine {
  readonly roleId: string;
  readonly values: readonly SealedValue[];
  readonly releases: readonly Release[];
}

export const loadPerson = (roleId: string): Promise<Mine> =>
  call(`/workspace/person/${encodeURIComponent(roleId)}/values`);

/**
 * Die eigenen Angaben aufmachen.
 *
 * Eine Hülle, die nicht aufgeht, wird übergangen: ein beschädigtes Feld darf
 * nicht die übrigen verstecken.
 */
export async function openMine(
  roleId: string, ring: Ring, values: readonly SealedValue[]
): Promise<Map<PersonField, string>> {
  const key = ring.keyOf(roleId);
  const out = new Map<PersonField, string>();

  for (const one of values) {
    try {
      out.set(one.field, await openText(key, valueAad(roleId, one.field), fromBase64Url(one.sealed)));
    } catch {
      // Nicht lesbar — aber die anderen schon.
    }
  }

  return out;
}

/* -- Schreiben -------------------------------------------------------------- */

/** Wohin die Berichtigung ging — und wohin nicht. */
export interface Refreshed {
  /** Bereiche, deren Abschrift jetzt wieder stimmt. */
  readonly refreshed: readonly string[];

  /** Bereiche, an die man nicht herankam. Namen, damit man es dem Menschen sagen kann. */
  readonly couldNot: readonly string[];
}

/** Was der Dienst zur Änderung sagt: wo diese Angabe schon liegt. */
interface Stale {
  readonly areaId: string;
  readonly areaName: string;
  readonly epoch: number;
}

/**
 * Eine Angabe setzen — und die Abschriften nachziehen.
 *
 * <b>Wer seine Nummer ändert, meint seine Nummer.</b> Nicht „meine Nummer hier
 * und die alte überall sonst". Die Angabe liegt an der Rolle des Menschen, und
 * jeder Bereich, dem er sie einmal gegeben hat, hält eine eigene Hülle davon.
 * Ohne diesen Schritt bliebe in jedem einzelnen davon der alte Stand stehen —
 * und niemand sähe es ihm an.
 *
 * <b>Der Dienst kann das nicht.</b> Eine Abschrift liegt unter dem
 * Epochenschlüssel eines Bereichs; er hat keinen. Er sagt nur, WELCHE Bereiche
 * betroffen sind — neu versiegelt wird hier, im Browser dessen, dem die Angabe
 * gehört.
 *
 * <b>Das gibt nichts Neues heraus.</b> Berichtigt wird nur, wo die Angabe schon
 * liegt. Ein Bereich, der sie nie bekommen hat, bekommt sie auch jetzt nicht;
 * dafür bleibt es bei `releaseValue`.
 *
 * <b>Zwei Wege an den Schlüssel, und der eigene zuerst.</b> Wer im Bereich
 * steht, hält den Epochenschlüssel selbst. Wer nur einmal etwas in ein
 * öffentliches Formular geschrieben hat, hält ihn nicht — dann der
 * offengelegte. Gibt es beides nicht, bleibt diese eine Abschrift alt, und das
 * wird BERICHTET statt verschwiegen: eine stumme Lücke wäre schlimmer als eine
 * genannte.
 */
export async function setValue(
  roleId: string, ring: Ring, field: PersonField, value: string
): Promise<Refreshed> {
  const text = value.trim();
  const sealed = await sealText(ring.keyOf(roleId), valueAad(roleId, field), text);

  const said = await call<{ stale?: readonly Stale[] }>(
    `/workspace/person/${encodeURIComponent(roleId)}/value`,
    { method: 'POST', body: JSON.stringify({ field, sealed: toBase64Url(sealed) }) });

  const refreshed: string[] = [];
  const couldNot: string[] = [];

  for (const where of said.stale ?? []) {
    try {
      await releaseValue(roleId, field, text, {
        areaId: where.areaId, areaKey: await epochKeyOf(ring, where), epoch: where.epoch
      });

      refreshed.push(where.areaName);
    } catch {
      /* Kein Schlüssel, oder der Bereich nimmt gerade nichts an. Die übrigen
         sind trotzdem berichtigt — und der Name steht in der Meldung. */
      couldNot.push(where.areaName);
    }
  }

  return { refreshed, couldNot };
}

/** Der eigene Epochenschlüssel, sonst der offengelegte, sonst gar keiner. */
async function epochKeyOf(ring: Ring, where: Stale): Promise<Uint8Array> {
  const held = (await myEpochKeys(ring, where.areaId)).get(where.epoch);
  if (held !== undefined) return held;

  const open = await loadPublicKey(where.areaId);

  if (open.epoch !== where.epoch) {
    throw new Error('Jawny klucz jest z innej epoki.');
  }

  return fromBase64Url(open.key);
}

/** Löschen — mitsamt allen Freigaben, denn sie hängen daran. */
export const forgetValue = (
  roleId: string, field: PersonField
): Promise<{ forgotten: boolean; withdrawn: number }> =>
  call(`/workspace/person/${encodeURIComponent(roleId)}/forget`, {
    method: 'POST',
    body: JSON.stringify({ field })
  });

/* -- Freigeben -------------------------------------------------------------- */

/**
 * EINE Angabe an EINEN Bereich.
 *
 * Der Wert wird hier neu versiegelt — unter dem Epochenschlüssel des
 * Empfängers. Wer eine Telefonnummer bekommen soll, bekommt genau die und nicht
 * den Geburtstag dazu.
 */
export async function releaseValue(
  roleId: string, field: PersonField, value: string,
  to: { areaId: string; areaKey: Uint8Array; epoch: number }
): Promise<void> {
  const sealed = await sealText(to.areaKey, releaseAad(roleId, field), value.trim());

  await call(`/workspace/person/${encodeURIComponent(roleId)}/release`, {
    method: 'POST',
    body: JSON.stringify({
      field, areaId: to.areaId, epoch: to.epoch, sealed: toBase64Url(sealed)
    })
  });
}

/**
 * Zurücknehmen.
 *
 * Die Zeile fällt. Was jemand gelesen hat, hat er gelesen — der Dienst sagt das
 * selbst mit, und die Oberfläche gibt es weiter, statt mehr zu versprechen, als
 * die Sache hergibt.
 */
export const withdrawValue = (
  roleId: string, field: PersonField, areaId: string
): Promise<{ withdrawn: boolean; note: string }> =>
  call(`/workspace/person/${encodeURIComponent(roleId)}/withdraw`, {
    method: 'POST',
    body: JSON.stringify({ field, areaId })
  });

/* -- Was ein Bereich bekommen hat ------------------------------------------- */

export interface Given {
  readonly roleId: string;
  readonly field: PersonField;
  readonly epoch: number;
  readonly sealed: string;
  readonly releasedAt: string;
}

export const loadGiven = (areaId: string): Promise<{ given: readonly Given[] }> =>
  call(`/workspace/area/${encodeURIComponent(areaId)}/people`);

/** Und aufmachen — mit dem Epochenschlüssel des Bereichs. */
export async function openGiven(
  given: readonly Given[], areaKey: Uint8Array
): Promise<Map<string, Map<PersonField, string>>> {
  const out = new Map<string, Map<PersonField, string>>();

  for (const one of given) {
    try {
      const text = await openText(
        areaKey, releaseAad(one.roleId, one.field), fromBase64Url(one.sealed));

      const row = out.get(one.roleId) ?? new Map<PersonField, string>();
      row.set(one.field, text);
      out.set(one.roleId, row);
    } catch {
      // Aus einer anderen Epoche, oder beschädigt. Die übrigen bleiben lesbar.
    }
  }

  return out;
}

File: backend/Api/Person.cs
using Kernel;
using Microsoft.Data.SqlClient;

namespace Api;

/// <summary>
/// Die Angaben eines Menschen — und wie EINE davon herausgeht.
///
/// <para>
/// <b>Das Einzige auf dieser Plattform, was einem MENSCHEN gehoert und keinem
/// Bereich.</b> Vorname, Nachname, Telefon, Geburtstag, Anschrift, E-Mail
/// liegen an seiner Rolle, einzeln versiegelt unter seinem Rollenschluessel.
/// Keine Pfarrei, keine Schule und kein Verein besitzt sie; sie werden
/// FREIGEGEBEN, und zwar einzeln.
/// </para>
///
/// <code>
///   person_value     das Original, unter dem Schluessel des Menschen
///   person_release   eine Abschrift, unter dem Epochenschluessel EINES Bereichs
/// </code>
///
/// <para>
/// <b>Warum eine Abschrift und nicht ein Verweis.</b> Ein Verweis hiesse, dass
/// der Bereich beim Lesen an den Schluessel des Menschen muesste — dann waere
/// die Freigabe ein Versprechen und keine Massnahme. So bekommt jeder Empfaenger
/// seine eigene Huelle, und wer eine Telefonnummer bekommen soll, bekommt genau
/// die und nicht den Geburtstag dazu.
/// </para>
///
/// <para>
/// <b>Zuruecknehmen heisst: die Zeile faellt.</b> Was jemand gelesen hat, hat
/// er gelesen; das kann keine Datenbank rueckgaengig machen, und diese tut auch
/// nicht so. Was aufhoert, ist der Zugriff auf das, was danach kommt.
/// </para>
///
/// <para>
/// Dass man nur freigeben kann, was man hat, steht nicht hier, sondern im
/// Fremdschluessel auf <c>(role_id, field)</c> — eine Regel, die die Datenbank
/// haelt, ist eine, die kein Aufruf umgeht.
/// </para>
/// </summary>
public static class Person
{
    /// <summary>
    /// Dieselben sieben wie <c>ck_person_value_field</c>.
    ///
    /// <para>
    /// Sie stehen hier ein zweites Mal, damit ein Tippfehler eine Meldung
    /// ergibt und keine Nummer aus der Datenbank. Laufen sie auseinander,
    /// lehnt die Datenbank ab, was dieser Dienst durchgelassen hat — deshalb
    /// nennt der Kommentar die Gegenstelle.
    /// </para>
    /// </summary>
    private static readonly string[] Fields =
        ["given_name", "surname", "phone", "born", "address", "email", "nickname"];

    public static void Map(WebApplication app)
    {
        app.MapGet("/workspace/person/{roleId:guid}/values", ValuesAsync);
        app.MapPost("/workspace/person/{roleId:guid}/value", SetAsync);
        app.MapPost("/workspace/person/{roleId:guid}/forget", ForgetAsync);

        app.MapPost("/workspace/person/{roleId:guid}/release", ReleaseAsync);
        app.MapPost("/workspace/person/{roleId:guid}/withdraw", WithdrawAsync);

        /*
         * Was ein BEREICH ueber Menschen bekommen hat. Die Gegenrichtung, und
         * sie liest nur, was ihr gegeben wurde — es gibt keinen Aufruf, der
         * vom Bereich aus in `person_value` hineinsieht.
         */
        app.MapGet("/workspace/area/{id:guid}/people", GivenAsync);
    }

    /* -- Die eigenen Angaben ------------------------------------------------ */

    private static async Task ValuesAsync(HttpContext ctx, Db db, Guid roleId)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        await using var connection = await db.OpenAsync(ctx.RequestAborted);
        if (!await MineAsync(ctx, connection, who.Value.AccountId, roleId)) return;

        var values = new List<object>();

        await using (var cmd = new SqlCommand(
            "SELECT field, value_sealed, updated_at FROM app.person_value WHERE role_id = @r ORDER BY field;",
            connection))
        {
            cmd.Parameters.AddWithValue("@r", roleId);

            await using var reader = await cmd.ExecuteReaderAsync(ctx.RequestAborted);
            while (await reader.ReadAsync(ctx.RequestAborted))
            {
                values.Add(new
                {
                    field = reader.GetString(0),
                    @sealed = Base64Url.Encode((byte[])reader[1]),
                    updatedAt = reader.GetDateTimeOffset(2)
                });
            }
        }

        /*
         * WOHIN etwas gegangen ist, gehoert neben das, was dasteht. Eine
         * Oberflaeche, die die Angaben zeigt und die Freigaben woanders, laesst
         * die eine Frage offen, die der Mensch wirklich hat: wer hat meine
         * Telefonnummer?
         */
        var releases = new List<object>();

        await using (var cmd = new SqlCommand("""
            SELECT p.field, p.area_id, a.name, p.epoch, p.released_at
            FROM app.person_release p
            JOIN app.area a ON a.id = p.area_id
            WHERE p.role_id = @r
            ORDER BY a.name, p.field;
            """, connection))
        {
            cmd.Parameters.AddWithValue("@r", roleId);

            await using var reader = await cmd.ExecuteReaderAsync(ctx.RequestAborted);
            while (await reader.ReadAsync(ctx.RequestAborted))
            {
                releases.Add(new
                {
                    field = reader.GetString(0),
                    areaId = Ids.ToText(reader.GetGuid(1)),
                    areaName = reader.GetString(2),
                    epoch = reader.GetInt32(3),
                    releasedAt = reader.GetDateTimeOffset(4)
                });
            }
        }

        await ctx.Response.WriteAsJsonAsync(new { roleId = Ids.ToText(roleId), values, releases });
    }

    public sealed record SetRequest(string Field, string Sealed);

    /// <summary>
    /// Eine Angabe setzen oder ersetzen.
    ///
    /// <para>
    /// <b>Der Dienst zieht die Abschriften nicht nach — er kann es nicht.</b>
    /// Eine Abschrift liegt unter dem Epochenschluessel eines Bereichs, und den
    /// hat er nicht. Was er kann, ist SAGEN, welche jetzt veraltet sind: die
    /// Antwort nennt jeden Bereich, der diese Angabe schon hat, samt Epoche.
    /// Neu versiegeln muss der Browser dessen, dem die Angabe gehoert.
    /// </para>
    ///
    /// <para>
    /// <b>Und er zieht sie nach, ohne zu fragen.</b> Wer seine Nummer aendert,
    /// meint seine Nummer — nicht „meine Nummer hier und die alte ueberall
    /// sonst". Die genannten Bereiche haben die Angabe bereits; sie zu
    /// berichtigen gibt nichts heraus, was sie nicht schon haetten. Was
    /// weiterhin niemand automatisch bekommt, ist eine Angabe, die er noch nie
    /// hatte — dafuer bleibt es bei <c>release</c>.
    /// </para>
    /// </summary>
    private static async Task SetAsync(HttpContext ctx, Db db, Guid roleId, SetRequest body)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        var field = (body.Field ?? string.Empty).Trim().ToLowerInvariant();

        if (!Fields.Contains(field))
        {
            await Fail(ctx, StatusCodes.Status400BadRequest,
                "Pole: imię, nazwisko, telefon, data urodzenia, adres, e-mail albo przezwisko.");
            return;
        }

        byte[] blob;
        try { blob = Base64Url.Decode(body.Sealed ?? string.Empty); }
        catch (FormatException)
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Nieczytelna zapieczętowana treść.");
            return;
        }

        if (blob.Length == 0)
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Puste pole nie jest zapieczętowane.");
            return;
        }

        await using var connection = await db.OpenAsync(ctx.RequestAborted);
        if (!await MineAsync(ctx, connection, who.Value.AccountId, roleId)) return;

        await using var cmd = new SqlCommand("""
            UPDATE app.person_value SET value_sealed = @blob, updated_at = @now
             WHERE role_id = @r AND field = @f;

            IF @@ROWCOUNT = 0
                INSERT INTO app.person_value (role_id, field, value_sealed, updated_at)
                VALUES (@r, @f, @blob, @now);
            """, connection);

        cmd.Parameters.AddWithValue("@r", roleId);
        cmd.Parameters.AddWithValue("@f", field);
        cmd.Parameters.AddWithValue("@blob", blob);
        cmd.Parameters.AddWithValue("@now", DateTimeOffset.UtcNow);

        await cmd.ExecuteNonQueryAsync(ctx.RequestAborted);

        /*
         * WO DIESE ANGABE SCHON LIEGT. Erst nach dem Schreiben gefragt: was
         * hier herauskommt, ist die Arbeitsliste fuer den Browser, und die
         * soll zu dem passen, was jetzt dasteht.
         *
         * Die laufende Epoche des Bereichs, nicht die der Abschrift — unter
         * einer abgeloesten Epoche neu zu versiegeln hiesse, die Berichtigung
         * dorthin zu legen, wo niemand mehr nachsieht.
         */
        var stale = new List<object>();

        await using (var ask = new SqlCommand("""
            SELECT r.area_id, a.name,
                   (SELECT MAX(e.epoch) FROM app.area_epoch e WHERE e.area_id = r.area_id)
            FROM app.person_release r
            JOIN app.area a ON a.id = r.area_id
            WHERE r.role_id = @r AND r.field = @f
            ORDER BY a.name;
            """, connection))
        {
            ask.Parameters.AddWithValue("@r", roleId);
            ask.Parameters.AddWithValue("@f", field);

            await using var reader = await ask.ExecuteReaderAsync(ctx.RequestAborted);
            while (await reader.ReadAsync(ctx.RequestAborted))
            {
                /* Ohne Epoche liesse sich nichts versiegeln; dann steht der
                   Bereich nicht auf der Liste, statt sie unbrauchbar zu machen. */
                if (reader.IsDBNull(2)) continue;

                stale.Add(new
                {
                    areaId = Ids.ToText(reader.GetGuid(0)),
                    areaName = reader.GetString(1),
                    epoch = reader.GetInt32(2)
                });
            }
        }

        await ctx.Response.WriteAsJsonAsync(new
        {
            roleId = Ids.ToText(roleId), field, saved = true, stale
        });
    }

    public sealed record ForgetRequest(string Field);

    /// <summary>
    /// Eine Angabe loeschen — mitsamt ihren Freigaben.
    ///
    /// <para>
    /// Die Freigaben MUESSEN mit: der Fremdschluessel zeigt auf die Angabe, und
    /// eine Abschrift, deren Original nicht mehr besteht, waere eine, die
    /// niemand mehr zurueckziehen kann.
    /// </para>
    /// </summary>
    private static async Task ForgetAsync(HttpContext ctx, Db db, Guid roleId, ForgetRequest body)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        var field = (body.Field ?? string.Empty).Trim().ToLowerInvariant();

        await using var connection = await db.OpenAsync(ctx.RequestAborted);
        if (!await MineAsync(ctx, connection, who.Value.AccountId, roleId)) return;

        await using var tx = (SqlTransaction)await connection.BeginTransactionAsync(ctx.RequestAborted);

        try
        {
            int withdrawn;

            await using (var drop = new SqlCommand(
                "DELETE FROM app.person_release WHERE role_id = @r AND field = @f;", connection, tx))
            {
                drop.Parameters.AddWithValue("@r", roleId);
                drop.Parameters.AddWithValue("@f", field);
                withdrawn = await drop.ExecuteNonQueryAsync(ctx.RequestAborted);
            }

            await using (var drop = new SqlCommand(
                "DELETE FROM app.person_value WHERE role_id = @r AND field = @f;", connection, tx))
            {
                drop.Parameters.AddWithValue("@r", roleId);
                drop.Parameters.AddWithValue("@f", field);
                await drop.ExecuteNonQueryAsync(ctx.RequestAborted);
            }

            await tx.CommitAsync(ctx.RequestAborted);
            await ctx.Response.WriteAsJsonAsync(new { field, forgotten = true, withdrawn });
        }
        catch
        {
            await tx.RollbackAsync(ctx.RequestAborted);
            throw;
        }
    }

    /* -- Freigeben ---------------------------------------------------------- */

    public sealed record ReleaseRequest(string Field, string AreaId, int Epoch, string Sealed);

    /// <summary>
    /// EINE Angabe an EINEN Bereich.
    ///
    /// <para>
    /// Der Browser hat sie unter dem Epochenschluessel dieses Bereichs neu
    /// versiegelt — er kennt ihn, weil er ihn oeffentlich holen konnte
    /// (<c>/area/{id}/key</c>) oder weil er ihn selbst haelt. Der Dienst legt
    /// die Huelle ab und kann sie nicht oeffnen.
    /// </para>
    ///
    /// <para>
    /// <b>Es wird NICHT geprueft, ob der Bereich darum gebeten hat.</b> Eine
    /// Freigabe ist eine Handlung des Menschen, keine Antwort auf eine
    /// Anforderung — und ein Bereich, der ungefragt etwas bekommt, hat nichts
    /// gewonnen, was er nicht ohnehin bekommen haette.
    /// </para>
    /// </summary>
    private static async Task ReleaseAsync(HttpContext ctx, Db db, Guid roleId, ReleaseRequest body)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        var field = (body.Field ?? string.Empty).Trim().ToLowerInvariant();

        if (!Fields.Contains(field))
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Takiego pola nie ma.");
            return;
        }

        if (!Guid.TryParse(body.AreaId, out var areaId) || body.Epoch < 1)
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Nieczytelny obszar albo epoka.");
            return;
        }

        byte[] blob;
        try { blob = Base64Url.Decode(body.Sealed ?? string.Empty); }
        catch (FormatException)
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Nieczytelna zapieczętowana treść.");
            return;
        }

        if (blob.Length == 0)
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Puste pole nie jest zapieczętowane.");
            return;
        }

        await using var connection = await db.OpenAsync(ctx.RequestAborted);
        if (!await MineAsync(ctx, connection, who.Value.AccountId, roleId)) return;

        await using var cmd = new SqlCommand("""
            UPDATE app.person_release
               SET epoch = @epoch, sealed_blob = @blob, released_at = @now
             WHERE role_id = @r AND field = @f AND area_id = @area;

            IF @@ROWCOUNT = 0
                INSERT INTO app.person_release (role_id, field, area_id, epoch, sealed_blob, released_at)
                VALUES (@r, @f, @area, @epoch, @blob, @now);
            """, connection);

        cmd.Parameters.AddWithValue("@r", roleId);
        cmd.Parameters.AddWithValue("@f", field);
        cmd.Parameters.AddWithValue("@area", areaId);
        cmd.Parameters.AddWithValue("@epoch", body.Epoch);
        cmd.Parameters.AddWithValue("@blob", blob);
        cmd.Parameters.AddWithValue("@now", DateTimeOffset.UtcNow);

        try
        {
            await cmd.ExecuteNonQueryAsync(ctx.RequestAborted);
        }
        catch (SqlException e) when (e.Number == 547)
        {
            /*
             * Der Fremdschluessel auf (role_id, field). Freigeben kann man nur,
             * was man hat — und das ist hier keine Meldung, sondern eine Regel
             * der Datenbank, die dieser Dienst gar nicht umgehen koennte.
             */
            await Fail(ctx, StatusCodes.Status409Conflict,
                "Najpierw wpisz to pole u siebie — nie można udostępnić czegoś, czego nie masz.");
            return;
        }

        await ctx.Response.WriteAsJsonAsync(new
        {
            roleId = Ids.ToText(roleId),
            field,
            areaId = Ids.ToText(areaId),
            released = true
        });
    }

    public sealed record WithdrawRequest(string Field, string AreaId);

    private static async Task WithdrawAsync(HttpContext ctx, Db db, Guid roleId, WithdrawRequest body)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        if (!Guid.TryParse(body.AreaId, out var areaId))
        {
            await Fail(ctx, StatusCodes.Status400BadRequest, "Nieczytelna kennung obszaru.");
            return;
        }

        await using var connection = await db.OpenAsync(ctx.RequestAborted);
        if (!await MineAsync(ctx, connection, who.Value.AccountId, roleId)) return;

        await using var cmd = new SqlCommand(
            "DELETE FROM app.person_release WHERE role_id = @r AND field = @f AND area_id = @area;",
            connection);

        cmd.Parameters.AddWithValue("@r", roleId);
        cmd.Parameters.AddWithValue("@f", (body.Field ?? string.Empty).Trim().ToLowerInvariant());
        cmd.Parameters.AddWithValue("@area", areaId);

        var gone = await cmd.ExecuteNonQueryAsync(ctx.RequestAborted) > 0;

        await ctx.Response.WriteAsJsonAsync(new
        {
            withdrawn = gone,

            /*
             * Ehrlich gesagt, und deshalb hier: was gelesen wurde, ist gelesen.
             * Eine Oberflaeche, die „zurueckgezogen" sagt und mehr verspricht,
             * als die Sache hergibt, waere die eigentliche Luege.
             */
            note = "Zapis usunięty. Co już odczytano, pozostaje odczytane."
        });
    }

    /* -- Was ein Bereich bekommen hat --------------------------------------- */

    private static async Task GivenAsync(HttpContext ctx, Db db, Guid id)
    {
        var who = await Auth.WhoAsync(ctx, db);
        if (who is null) { ctx.Response.StatusCode = StatusCodes.Status401Unauthorized; return; }

        await using var connection = await db.OpenAsync(ctx.RequestAborted);

        if (!await Area.MayAsync(connection, who.Value.AccountId, id, Capability.Read, ctx.RequestAborted))
        {
            await Fail(ctx, StatusCodes.Status404NotFound, "Takiego obszaru nie ma.");
            return;
        }

        await using var cmd = new SqlCommand("""
            SELECT role_id, field, epoch, sealed_blob, released_at
            FROM app.person_release
            WHERE area_id = @area
            ORDER BY role_id, field;
            """, connection);

        cmd.Parameters.AddWithValue("@area", id);

        var given = new List<object>();

        await using var reader = await cmd.ExecuteReaderAsync(ctx.RequestAborted);
        while (await reader.ReadAsync(ctx.RequestAborted))
        {
            given.Add(new
            {
                roleId = Ids.ToText(reader.GetGuid(0)),
                field = reader.GetString(1),
                epoch = reader.GetInt32(2),
                @sealed = Base64Url.Encode((byte[])reader[3]),
                releasedAt = reader.GetDateTimeOffset(4)
            });
        }

        await ctx.Response.WriteAsJsonAsync(new { areaId = Ids.ToText(id), given });
    }

    /* -- Gemeinsames -------------------------------------------------------- */

    /// <summary>
    /// Gehoert diese Rolle DIESEM Konto?
    ///
    /// <para>
    /// Die einzige Frage, die hier zaehlt. Ein Bereich, ein Zertifikat, eine
    /// Mitgliedschaft geben keinen Zugriff auf die Angaben eines Menschen —
    /// sie gehoeren ihm, nicht seiner Umgebung.
    /// </para>
    /// </summary>
    private static async Task<bool> MineAsync(
        HttpContext ctx, SqlConnection connection, Guid accountId, Guid roleId)
    {
        var mine = await Workspace.RolesOfAsync(connection, accountId, ctx.RequestAborted);
        if (mine.Any(r => r.Id == roleId)) return true;

        await Fail(ctx, StatusCodes.Status403Forbidden, "To nie jest Twoja rola.");
        return false;
    }

    private static Task Fail(HttpContext ctx, int status, string message)
    {
        ctx.Response.StatusCode = status;
        return ctx.Response.WriteAsJsonAsync(new { error = message });
    }
}
